src/rfc/rebuild_invocation_args.php

```diff
@@ -0,0 +1,185 @@
+      ->setProtected()
+      ->setBody('return $this->arg_map;'),
+  );
+
+  foreach ($args as $arg) {
+    $method = $cg->codegenMethod($arg['func_name'])
+      ->setPublic()
+      ->setIsFinal()
+      ->setReturnType('this');
+    $right_hand_side = 'null';
+    if ($arg['takes_arg']) {
+      $method->addParameter('string $value');
+      $right_hand_side = '$value';
+    }
+    $method->setBody(
+      '$this->arg_map["'.
+      $arg['cli_name'].
+      '"] = '.
+      $right_hand_side.
+      ";\nreturn \$this;",
+    );
+    if (Shapes::idx($arg, 'description') != null) {
+      $method->setDocBlock($arg['description']);
+    }
+    $trait->addMethod($method);
+  }
+  $cg->codegenFile(__DIR__.'/ShipItArgumentsTrait.php')
+    ->setDocBlock("Copyright (c) Facebook, Inc. and its affiliates.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.")
+    ->setNamespace("Facebook\ShipIt")
+    ->addTrait($trait)
+    ->save();
+}
+
+<<__EntryPoint>>
+async function run_test_main(): Awaitable<void> {
+  require_once(\dirname(\dirname(__DIR__)).'/vendor/autoload.hack');
+  \Facebook\AutoloadMap\initialize();
+
+  $phases = vec[
+    new ShipItCleanPhase(ShipItRepoSide::SOURCE),
+    new ShipItCleanPhase(ShipItRepoSide::DESTINATION),
+    new ShipItCreateNewRepoPhase($x ==> $x, shape('name' => '', 'email' => '')),
+    new ShipItDeleteCorruptedRepoPhase(ShipItRepoSide::SOURCE),
+    new ShipItDeleteCorruptedRepoPhase(ShipItRepoSide::DESTINATION),
+    new ShipItFilterSanityCheckPhase($x ==> $x, vec[]),
+    new ShipItGitHubInitPhase(
+      '',
+      '',
+      ShipItRepoSide::SOURCE,
+      ShipItTransport::HTTPS,
+      DemoGitHubUtils::class,
+    ),
+    new ShipItGitHubInitPhase(
+      '',
+      '',
+      ShipItRepoSide::DESTINATION,
+      ShipItTransport::HTTPS,
+      DemoGitHubUtils::class,
+    ),
+    new ShipItPullPhase(ShipItRepoSide::SOURCE),
+    new ShipItPullPhase(ShipItRepoSide::DESTINATION),
+    new ShipItPushLfsPhase(
+      ShipItRepoSide::SOURCE,
+      '',
+      '',
+      false,
+      DemoGitHubUtils::class,
+    ),
+    new ShipItPushLfsPhase(
+      ShipItRepoSide::DESTINATION,
+      '',
+      '',
+      false,
+      DemoGitHubUtils::class,
+    ),
+    new ShipItPushPhase(),
+    new ShipItSaveConfigPhase('', ''),
+    new ShipItVerifyRepoPhase($x ==> $x),
+  ];
+
+  $args = (
+    new ShipItPhaseRunner(
+      new ShipItBaseConfig('', '', '', keyset['.']),
+      $phases,
+      new ShipItCLIArgumentParser(),
+    )
+  )->getCLIArguments();
+
+  $function_args = Vec\map($args, ($arg) ==> getFunctionArg($arg))
+    |> Vec\filter_nulls($$)
+    |> Vec\sort($$);
+
+  $unique_function_args = vec[];
+  foreach ($function_args as $arg) {
+    if (C\last($unique_function_args) == $arg) {
+      continue;
+    }
+    $unique_function_args[] = $arg;
+  }
+
+  generateArgumentClass($unique_function_args);
```

Comment:
Can this be a public method that takes in an arbitrary phases list? That way, the different entry points can each have their generated args.